tests/common/ixia/common_helpers.py

```diff
@@ -200,11 +200,182 @@ def pfc_class_enable_vector(prio_list):
 
     Returns:
-        Return class-enable vector 
+        Return class-enable vector
     """
     vector = 0
 
     for p in prio_list:
         vector += (2**p)
-     
+
     return "{:x}".format(vector)
+
+def get_wred_profiles(host_ans):
+    """
+    Get all the WRED/ECN profiles of a SONiC switch
+
+    Args:
+        host_ans: Ansible host instance of the device
+
+    Returns:
+        WRED/ECN profiles (dictionary) or None.
+        Example format is given below:
+
+        {
+            u'AZURE_LOSSLESS': {
+                u'ecn': u'ecn_all',
+                u'green_drop_probability': u'5',
+                u'green_max_threshold': u'2097152',
+                u'green_min_threshold': u'250000',
+                u'red_drop_probability': u'5',
+                u'red_max_threshold': u'2097152',
+                u'red_min_threshold': u'1048576',
+                u'wred_green_enable': u'true',
+                u'wred_red_enable': u'true',
+                u'wred_yellow_enable': u'true',
+                u'yellow_drop_probability': u'5',
+                u'yellow_max_threshold': u'2097152',
+                u'yellow_min_threshold': u'1048576'
+            }
+        }
+    """
+    config_facts = host_ans.config_facts(host=host_ans.hostname,
+                                         source="running")['ansible_facts']
+
+    if "WRED_PROFILE" in config_facts.keys():
+        return config_facts['WRED_PROFILE']
+    else:
+        return None
+
+def config_wred(host_ans, kmin, kmax, pmax, profile=None):
+    """
+    Config a WRED/ECN profile of a SONiC switch
+
+    Args:
+        host_ans: Ansible host instance of the device
+        kmin (int): RED/ECN minimum threshold in bytes
+        kmax (int): RED/ECN maximum threshold in bytes
+        pmax (int): RED/ECN maximum marking probability in percentage
+        profile (str): name of profile to configure (None means any profile)
+
+    Returns:
+        If configuration succeeds (bool)
+    """
+
+    if not isinstance(kmin, int) or \
+       not isinstance(kmax, int) or \
+       not isinstance(pmax, int):
+        return False
+
+    if kmin < 0 or kmax < 0 or pmax < 0 or pmax > 100 or kmin > kmax:
+        return False
+
+    profiles = get_wred_profiles(host_ans)
+    """ Cannot find any WRED/ECN profiles """
+    if profiles is None:
+        return False
+
+    """ Cannot find the profile to configure at the device """
+    if profile is not None and profile not in profiles:
+        return False
+
+    for p in profiles:
+        """ This is not the profile to configure """
+        if profile is not None and profile != p:
+            continue
+
+        kmin_old = int(profiles[p]['green_min_threshold'])
+        kmax_old = int(profiles[p]['green_max_threshold'])
+
+        if kmin_old > kmax_old:
+            return False
+
+        """ Ensure that Kmin is no larger than Kmax during the update """
+        if kmin > kmin_old:
```

Comment:
why is this check needed? you have already checked earlier that kmin < kmax. so why do you need to set in this order

Reply:
This is a bug of current SONiC implementation. We need to ensure that Kmin is always smaller than Kmax during the update. For example, if we want to increase Kmin/Kmax from 100KB/200KB to 300KB/400KB, we need to increase Kmax first and then Kmin. Otherwise, Kmin will be larger than Kmax during the update, making swss collapse.

Reply:
I will check this one